### DISCUSSION TITLE
Best Approach for a Self-Contained .NET Application?

### DISCUSSION THREAD

**Discussion by eajhnsn1** [wixtoolset/issues#7403] [Questions] [answered]
I have a pair of small applications for which I'm using Wix 4 to install.  These are .NET 6 applications which I want installed as Self Contained applications.

Is there a better approach to this than adding a `BeforeBuild` msbuild target to the .wixproj which manually calls `dotnet publish .... --self-contained` and harvesting the resulting output?

**Comment by chrpai** [comment]
Here's what I find works really well for me.   I create a repos with an Application\Application.sln and Installer\Installer.sln.  My projects are like Application\Project\Project.csproj.

I create a FolderPublish Profile that looks like this.

FolderProfile.pubxml

```
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project>
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>x64</Platform>
    <PublishDir>..\..\Installer\Deploy</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <_TargetId>Folder</_TargetId>
    <TargetFramework>net7.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishSingleFile>false</PublishSingleFile>
    <PublishReadyToRun>false</PublishReadyToRun>
    <PublishTrimmed>false</PublishTrimmed>
  </PropertyGroup>
</Project>
```

In he Project.csproj I added property   ` <PublishProfile>FolderProfile</PublishProfile>`

In my build I simply run:

`dotnet publish Application\Application.sln -c Release /p:Version=1.2.3.4`

This places everything I need in the Installer\Deploy folder.  You can exclude stuff like .PDBs. createdump.exe and if you see something like mscordaccore_amd64_amd64_7.0.523.17405.dll you don't want that either.  You only need mscordaccore.dll.


From there I use my IsWiX project templates to create my WiX v4 files, drag drop the files in from the deploy folder, create a shortcut and build it all using:

dotnet build $slnFile -c Release /p:MSIProductVersion=1.2.3 -o OUTPUT_ARTIFACTS_PATH

It's super easy these days.

**Comment by robmen** [comment, marked as answer]
Try adding `Publish='true'` to the `ProjectReference` in your .wixproj to your .csproj. So in "My.wixproj":

```xml
<ItemGroup>
  <ProjectReference Include="..\path\to\Your.csproj" Publish="true" />
</ItemGroup>
```

Then you can use the automatically added bindpath `!(bindpath.Your)` in your `Source` attribute to reference the files.

**Reply by eajhnsn1** [reply]
Fantastic, this is exactly what I was looking for.  Thank you Rob!

**Reply by maroonedsia** [reply]
It seems just publishing the project is not good enough, unless the used publish profile has the `<SelfContained>true</SelfContained>`... then my question was how to pick the publish profile for the build, and it seems chrpai's answer includes a clue. If I'm missing anything here, please let me know. Thanks! 👍

**Comment by maroonedsia** [comment]
Here is my solution, it's a combination of @robmen and @chrpai answers:

1. Create a publish profile for the project you want to publish (this is the project that is going to be referenced in the wixproj). When you right-click the project in the solution explorer and choose 'Publish', VS creates the `.pubxml` file in the `\Properties\PublishProfiles\` folder for you. NOTE: the default visual studio .gitignore ignores publish profiles, so if you want to check in this file, you need to remove the ignore rule. Make sure your publish profile includes `<SelfContained>true</SelfContained>`
2. Add this line to the to-be-published .csproj file's PropertyGroup: `<PublishProfile>[Your_Profile_Name]</PublishProfile>`
3. Add the `Publish="true"` to the `<ProjectReference>` of the .wixproj file.
